BUG: Dense ranking with percent now uses 100% basis

rouzazari opened the pull request:
- `DataFrame.rank()` and `Series.rank()` when `method='dense'` and
  `pct=True` now scales to 100%.

See #15630

 - [ ] closes #15630 
 - [ ] tests added / passed
 - [X] passes ``git diff upstream/master | flake8 --diff``
 - [X] whatsnew entry

rouzazari committed:
BUG: Dense ranking with percent now uses 100% basis

- `DataFrame.rank()` and `Series.rank()` when `method='dense'` and
  `pct=True` now scales to 100%.

See #15630

rouzazari committed:
Fixed typo in test when fixing PEP8/Flake8

jreback commented:
FYI #15640 (can be done after this one). Or before if you are so inclined.

jreback commented on pandas/tests/test_stats.py:
```diff
@@ -108,6 +108,31 @@ def test_rank_dense_method(self):
                 expected = Series(exp).astype(result.dtype)
                 assert_series_equal(result, expected)
 
+        # GH15630, pct should be on 100% basis even when method='dense'
```
make this a parametrized tests (needs to be a separate function, not in this class)

jreback reviewed (changes requested):
I think we also need to check for method='min'? (and others)?

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -855,3 +855,4 @@ Bug Fixes
 - Bug in ``pd.melt()`` where passing a tuple value for ``value_vars`` caused a ``TypeError`` (:issue:`15348`)
 - Bug in ``.eval()`` which caused multiline evals to fail with local variables not on the first line (:issue:`15342`)
 - Bug in ``pd.read_msgpack`` which did not allow to load dataframe with an index of type ``CategoricalIndex`` (:issue:`15487`)
```
I think we need a sub-section on this. To show an example case of the before and after.

rouzazari commented:
Thanks, @jreback. Agreed that tests should be split out a bit and we should be testing other ranking `method` too. May take a bit to work through but happy to tackle as much as I can.

jreback commented:
@rouzazari great. as I said, ok to leave tests where they are and split after / or do independent PR to split before (and rebase this one).

jreback commented:
ok, pls rebase (after https://github.com/pandas-dev/pandas/issues/15640) and we can see where this is

jreback commented:
can you update

rouzazari commented:
Yes. Been delayed but hoping to complete this week. Will ping when I have updated.

jreback commented:
np, thanks.

rouzazari committed:
Rebase and parametrize tests

1. Moved tests to series-/frame-level test_rank (from test_stats) per #15658.
2. Parametrize `series/test_rank` for `pct=True`.

rouzazari commented:
I rebased and parametrized the `Series` test, but I am still researching `method='min'` and others.

Also, we will still need a sub-section in `whatsnew` to show before/after examples

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -1113,3 +1113,4 @@ Other
 - Compat for 32-bit platforms for ``.qcut/cut``; bins will now be ``int64`` dtype (:issue:`14866`)
 - Bug in interactions with ``Qt`` when a ``QtApplication`` already exists (:issue:`14372`)
 - Avoid use of ``np.finfo()`` during ``import pandas`` removed to mitigate deadlock on Python GIL misuse (:issue:`14641`)
```
can you move to 0.21.0

jreback replied:
to the numeric section of bug fixes. I thin this is ok as a bug fix, though if you think as a user it would warrant a bigger mention, a sub-section would be ok too.

jreback commented on pandas/tests/series/test_rank.py:
```diff
@@ -322,3 +322,23 @@ def test_rank_object_bug(self):
+    ([-5, -4, -3, -2, -1], [1. / 5, 2. / 5, 3. / 5, 4. / 5, 5. / 5])])
+def test_rank_pct(dtype, ser, exp):
+        # GH15630, pct should be on 100% basis even when method='dense'
+        # TODO: add other methods (i.e. 'average', 'min', 'max', 'first')
```
can you add the other methods

jreback reviewed (approved):
looks like just some minor comments. pls ping when ready.

jreback commented:
can you update

rouzazari commented:
Yes. I've been a bit delayed, but working on a solution over the next couple days. Thanks for the comments above. I'll ping soon.

rouzazari committed:
Parameterize all rank methods for pct=True

- Added frame and series tests for all methods
- method='dense' remains the only change method

rouzazari committed:
Moved whatsnew to v.0.21.0

rouzazari commented:
I have rebased and added tests for all ``method``s with ``pct=True``. 

Please note that tests for all other methods assumes the existing calculation is working as intended. This bug fix only changes ``method=dense``; however, this is a good starting point should we decide to change ``pct=True`` overall.

I'd like to leave this as a bug fix in the docs for now until we fully consider the overall idea of ``pct=True`` as discussed in #15630. After we have an understanding of whether ``pct=True`` is always based on 100% or if this only applies to ``method=dense``, then we can add a sub-section on how ``pct=True`` is changing (if at all).

Please let me know your thoughts/comments, if any.

Will ping when green. Thanks

jreback commented:
@dsm054 can you have a look here.

gfyoung commented:
@rouzazari : Looks like you have a merge conflict in the `whatsnew`.

@jreback , @dsm054 : What are your thoughts on this one?  As it got approved already, and no comments have been made for nearly two months, I would think this is good to go.

jreback commented:
this looks pretty reasonable. I would expand the whatsnew to show how this is different than before (as its a numerical change, a more complete entry highlites it a bit more)

gfyoung commented:
@rouzazari : Would you be able to address @jreback 's suggestion for the `whatsnew` ?

dsm054 commented:
In the original ticket, we never got around to agreeing that we should fix other affected methods as well.  I haven't had my morning coffee yet so I could easily be missing something, but it looks like we're actually now testing that the behaviour of the others is _wrong_ in the same way that dense was wrong before this.

This work is definitely important enough to do what's needed to get into v0.21.0 IMHO.  Whether we deal with the other cases in this PR as well or open up another PR to handle it doesn't matter to me, but I think those should go in as well before the next release.

rouzazari commented:
Glad to see this come back to life. Thanks for comments all.

@dsm054 is absolutely correct that the new tests for ranking methods (other than dense ranking) are testing the expected results given the existing code. I am not sure what the correct results are for each ranking method, but I wanted to make sure we have tests for when we fix the ranking methods. 

If you can help me determine the correct results for each ranking method, then I can update the pull request, or we can close as is and fix in a separate pull request.

@gfyoung: I'm happy to add jreback's whatsnew suggestion. Will update soon.

jreback commented:
@rouzazari why don't you mark those 'suspect' ones as xfails, then we can address them one-by-one (in followups).

jreback commented:
can you rebase and add xfails for those suspect ones

jreback commented:
can you rebase

jreback commented:
can you rebase

jreback commented:
can you rebase

rouzazari commented:
Yes, sorry for the delay. I will rebase and add xfails in the coming week.

jreback commented:
thanks @rouzazari

auvipy commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -116,6 +116,7 @@ Reshaping
 
 Numeric
 ^^^^^^^
+- Bug in ``DataFrame.rank()`` and ``Series.rank()`` when ``method='dense'`` and ``pct=True`` (:issue:`15630`)
```
move this to 0.23.0.txt

auvipy reviewed (commented):
and plz rebase

jreback commented:
closing as stale. ping if you'd like to continue.

gfyoung commented:
@jreback : I'd like to continue this 😄

gfyoung commented:
@jreback : PTAL.  Was there anything that was needed before merging?

gfyoung commented:
@jreback : Friendly ping.

jreback commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -900,6 +900,7 @@ Offsets
 
 Numeric
 ^^^^^^^
+- Bug in ``DataFrame.rank()`` and ``Series.rank()`` when ``method='dense'`` and ``pct=True`` (:issue:`15630`)
```
can you make these ``:func:`` refs, can you specify a little more what the bug is/was.

jreback commented on pandas/tests/frame/test_rank.py:
```diff
@@ -266,3 +266,56 @@ def _check2d(df, expected, method='average', axis=0):
                 continue
             frame = df if dtype is None else df.astype(dtype)
```
OT: happy to clean these tests up above / parameterize and such. (next PR).

jreback commented on pandas/tests/frame/test_rank.py:
```diff
@@ -266,3 +266,56 @@ def _check2d(df, expected, method='average', axis=0):
+
+
+# GH15630, pct should be on 100% basis when method='dense'
+@pytest.mark.parametrize('frame, exp', [
```
can you make frame a fixture, its seems the same for each of these test functions. just the results are diff.

jreback replied:
or maybe better is to inline it inside the function and use a single parameterize on the ``method, exp``

gfyoung replied:
I like your second idea better.

jreback commented on pandas/tests/series/test_rank.py:
```diff
@@ -376,3 +376,96 @@ def test_rank_modify_inplace(self):
+
+# GH15630, pct should be on 100% basis when method='dense'
+
+@pytest.mark.parametrize('dtype', ['O', 'f8', 'i8'])
```
might consider putting these in a single parameterize as well (though not as clear cut as the frame case where you can share the input frame)

gfyoung replied:
Yeah...I can't think of a clean way to do this at the moment.  I'll address the other changes first, and then maybe we'll see if any inspiration strikes.  Otherwise, I would leave it alone.

gfyoung committed:
MAINT: Address reviewer comments

gfyoung commented:
Travis failure is also on <a href="https://travis-ci.org/pandas-dev/pandas/jobs/350554653">master</a>.  Seems to be happening on <a href="https://circleci.com/gh/pandas-dev/pandas/11139#tests/containers/2">Circle</a> as well.

jreback commented:
can you rebase on master.

gfyoung commented:
> can you rebase on master.

Done.

gfyoung commented:
@jreback : Much better!  All is green again.  PTAL.

jreback commented:
thanks @rouzazari and @gfyoung !
